### PR TITLE
Jobs

### PR DESCRIPTION
**Pull request by jinroh**
This PR add a small abstraction layer to define the basis of our job system as well as an implementation of a memory based job queues targeting self-hosting audience.

### CODE REVIEW

File: pkg/jobs/jobs.go [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+
+func makeID() string {
+	const (
+		idLen         = 16
+		letters       = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ"
+		letterIdxMask = 1<<6 - 1 // 6 bits to represent a letter index
+	)
+	b := make([]byte, idLen)
+	for i := 0; i < idLen; {
+		if idx := int(rand.Int63() & letterIdxMask); idx < len(letters) {
```
Comment (nono) [inline pkg/jobs/jobs.go]: And shouldn't we run `rand.Seed` to avoid having the same ids when the cozy-stack is restarted?

File: pkg/jobs/jobs.go [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+	// GobEncoding is a Gob encoding message type
+	GobEncoding = "gob"
+)
+
+type (
+	// Queue interface is used to represent a asynchronous queue of jobs from
+	// which it is possible to enqueue and consume jobs.
+	Queue interface {
+		Enqueue(*Job) error
+		Consume() (bool, *Job, error)
```
Comment (nono) [inline pkg/jobs/jobs.go]: What does the `bool` in the return mean?

File: pkg/instance/instance.go [hunk 3]
```diff
@@ -123,6 +124,11 @@ func (i *Instance) FS() afero.Fs {
 	return i.storage
 }
 
+// JobsBroker returns the jobs broker associated with the instance
+func (i *Instance) JobsBroker() (jobs.Broker, error) {
```
Comment (jinroh) [inline pkg/instance/instance.go]: I don't know.. For the case of the in-memory queue I would tend to agree but for the case of an external queue this getter may initialize a connection with an external system.

Although I agree these are problems we may not want to resolve yet...

File: pkg/jobs/jobs.go (hunk 2 above)
Comment (jinroh) [inline pkg/jobs/jobs.go]: To alert the consumer that the queue is closed. Might not be really idiomatic though.

File: pkg/jobs/jobs.go [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+		Foreach(func(*Job) error) error
```
Comment (jinroh) [inline pkg/jobs/jobs.go]: I'll remove it. It is indeed of no use for now.

In the jobs doc I documented the possibility for a user to fetch all the element of the queue but I'm not sure we really want to allow that.

File: pkg/jobs/jobs.go [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+		Len() int
+		Close()
+	}
+
+	// Broker interface is used to represent a job broker associated to a
+	// particular domain from which it is possible to create jobs that a pushed
+	// in the job system.
```
Comment (jinroh) [inline pkg/jobs/jobs.go]: No. However for the case of the memory queue it is the case, but it is particular.

I could add an explicit interface representing the responsibility to start the workers... but I'm not sure how to do it for now.

File: pkg/jobs/jobs.go [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+
+	// JobRequest struct is used to represent a new job request.
+	JobRequest struct {
+		WorkerType string
+		Message    *Message
+	}
+)
+
+// NewMessage returns a new Message encoded in the specified format.
+func NewMessage(enc string, data interface{}) (*Message, error) {
```
Comment (jinroh) [inline pkg/jobs/jobs.go]: It is not used for now. Should be used by triggers.

File: pkg/jobs/jobs.go [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+	var b []byte
+	var err error
+	switch enc {
+	case JSONEncoding:
+		b, err = json.Marshal(data)
+	case GobEncoding:
```
Comment (jinroh) [inline pkg/jobs/jobs.go]: Gob is a bit more compact and could be useful for long payload. If you think it is of no use, I'll remove it.

File: pkg/jobs/jobs.go (hunk 1 above)
Comment (jinroh) [inline pkg/jobs/jobs.go]: I did not know the seed was always the same on the global random source. I'll fix that.

File: pkg/instance/instance.go (hunk 3 above)
Comment (nono) [inline pkg/instance/instance.go]: Yes, we can add that later

File: pkg/jobs/jobs.go (hunk 2 above)
Comment (nono) [inline pkg/jobs/jobs.go]: At least, you should add a comment to explain that

File: pkg/jobs/jobs.go (hunk 4 above)
Comment (nono) [inline pkg/jobs/jobs.go]: The only use case I think about this is debugging. As a developer of a client-side applications, I may want to see what are the jobs in a queue. But, I'm not sure about it.

File: pkg/jobs/jobs.go (hunk 5 above)
Comment (nono) [inline pkg/jobs/jobs.go]: Without adding an explicit interface, a comment can be enough

File: pkg/jobs/jobs.go (hunk 1 above)
Comment (nono) [inline pkg/jobs/jobs.go]: Not sure, but I think it is the case

File: pkg/jobs/jobs.go (hunk 7 above)
Comment (nono) [inline pkg/jobs/jobs.go]: I think it is a bit prematured to add this

File: pkg/jobs/jobs.go [hunk 8]
```diff
@@ -113,16 +113,15 @@ func makeQueueName(domain, workerType string) string {
 
 func makeID() string {
 	const (
-		idLen         = 16
-		letters       = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ"
-		letterIdxMask = 1<<6 - 1 // 6 bits to represent a letter index
+		idLen   = 16
+		letters = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ"
 	)
+	r := rand.New(rand.NewSource(time.Now().UTC().UnixNano()))
```
Comment (nono) [inline pkg/jobs/jobs.go]: I'd have created a new source for the whole jobs package, not one per id. But it should work too

**Comment by jinroh** [comment]
Squashed into one commit

File: pkg/jobs/jobs.go [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+	default:
+		return ErrUnknownMessageType
+	}
+}
+
+func makeQueueName(domain, workerType string) string {
+	return fmt.Sprintf("%s/%s", domain, workerType)
+}
+
+func makeID() string {
```
Comment (aenario) [inline pkg/jobs/jobs.go]: may be we could move this function into a separate package (crypto.NonCryptoRandomID ?)

Comment (jinroh) [inline pkg/jobs/jobs.go]: some kind of `utils` package for instance ?

**Pull request merged**